src/resources/ble_esp32.cc

```diff
@@ -1403,64 +1473,99 @@ PRIMITIVE(init) {
   ByteArray* proxy = process->object_heap()->allocate_proxy();
   if (proxy == null) FAIL(ALLOCATION_FAILED);
 
+  BleEventSource* event_source = BleEventSource::instance();
+
+  BleResourceGroup* group = _new BleResourceGroup(process, event_source);
+  if (!group) FAIL(MALLOC_FAILED);
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(create_adapter) {
+  ARGS(BleResourceGroup, group)
+  // Note that we don't take the BLE lock yet.
+  // We are setting the lock in this function.
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) FAIL(ALLOCATION_FAILED);
+
   int id = ble_pool.any();
   if (id == kInvalidBle) FAIL(ALREADY_IN_USE);
 
-  BleEventSource* ble = BleEventSource::instance();
-
-  Mutex* mutex = OS::allocate_mutex(0, "BLE");
-  if (!mutex) {
+  ble_mutex = OS::allocate_mutex(0, "BLE");
+  if (!ble_mutex) {
     ble_pool.put(id);
     FAIL(MALLOC_FAILED);
   }
 
-  ble_hs_cfg.sync_cb = ble_on_sync;
+  // We can already set the callback, even though the adapter hasn't been created
+  // yet.
+  // The Adapter starts the NimBLE thread, so there won't be any callback before
+  // the adapter was created, and its instance_ field has been set.
+  ble_hs_cfg.sync_cb = BleAdapterResource::on_sync;
 
   // It is important to call nimble_port_init before creating the resource group, as the
   // resource group constructor starts the nimble background thread that uses
   // structures initialize by the init function.
   nimble_port_init();
 
-  auto group = _new BleResourceGroup(process, ble, id, mutex);
-  if (!group) {
-    OS::dispose(mutex);
-    ble_pool.put(id);
+  auto adapter = _new BleAdapterResource(group, id);
```

Comment:
It feels unfortunate to have to deinit nimble because of an allocation failure. Can we delay that now we're not creating the group here?

Reply:
I'm now calling the 'init' in the constructor of the Adapter.
done.